Moonscraper Chart Editor/Assets/Scripts/Game/Charts/IO/Midi/MidReader.cs

```diff
@@ -1122,5 +1238,69 @@ static void ProcessNoteOnEventAsFlagToggle(in EventProcessParams eventProcessPar
                 }
             }
         }
+
+        static void ProcessTextEventPairAsStarpower(in EventProcessParams eventProcessParams, string startText, string endText, Starpower.Flags flags = Starpower.Flags.None)
+        {
+            foreach (Song.Difficulty difficulty in EnumX<Song.Difficulty>.Values)
+            {
+                var song = eventProcessParams.song;
+                var instrument = eventProcessParams.instrument;
+                var chart = song.GetChart(instrument, difficulty);
+
+                // Retrieve start and end events
+                var startEvents = new List<ChartEvent>();
+                var endEvents = new List<ChartEvent>();
+                for (int i = 0; i < chart.events.Count; ++i)
+                {
+                    var textEvent = chart.events[i];
+                    if (textEvent.eventName == startText)
+                    {
+                        startEvents.Add(textEvent);
+                    }
+                    else if (textEvent.eventName == endText)
+                    {
+                        endEvents.Add(textEvent);
+                    }
+                }
+
+                // Don't process if there aren't the same number of start and end events
+                if (startEvents.Count != endEvents.Count)
+                {
+                    Debug.LogWarning($"Mismatch between start and end event counts on {difficulty} {instrument}. Cannont continue safely, skipping.");
+                    return;
+                }
+
+                // Pair together start and end events
+                for (int endIndex = 0; endIndex < endEvents.Count; ++endIndex)
+                {
```

Comment:
I think I just copied this from the new SysEx code when I did this initially, over there it has to keep track of multiple events being active at once. (or it was the other way around, one of these was copied after the other lol)

Regardless, yeah it would definitely be simpler to do it linearly.